echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/impl/MissedPipelineTriggerCompensationJob.groovy

```diff
@@ -0,0 +1,210 @@
+        }
+    }
+  }
+
+  void onPipelineCacheAwait(long tick) {
+    log.info('Waiting for pipeline cache to fill')
+  }
+
+  void onPipelineCacheError(Throwable t) {
+    log.error("Error waiting for pipeline cache: ${t.message}")
+  }
+
+  void triggerMissedExecutions(List<Pipeline> pipelines) {
+    log.info('Looking for missed pipeline executions from cron triggers')
+
+    pipelines = pipelines.findAll { !it.disabled }
+    List<Trigger> triggers = getEnabledCronTriggers(pipelines)
+
+    List<String> ids = getPipelineConfigIds(pipelines, triggers)
+    orcaService.getLatestPipelineExecutions(ids)
+      .subscribe({
+        onOrcaResponse(it, pipelines, triggers)
+      }, { onOrcaError(it) })
+  }
+
+  void onOrcaResponse(Collection<PipelineResponse> response, List<Pipeline> pipelines, List<Trigger> triggers) {
+    triggers.each { trigger ->
+      Pipeline pipeline = pipelines.find { it.triggers.contains(trigger) }
+      List<Date> executions = response.findAll { it.pipelineConfigId == pipeline.id }.collect {
+        // A null value is valid; a pipeline that hasn't started won't get re-triggered.
+        it.startTime != null ? new Date(it.startTime) : null
+      }
+
+      if (executions.size() == 0) {
+        return
+      }
+
+      def lastExecution = executions.first()
+      if (lastExecution == null) {
+        return
+      }
+
+      def expr = new CronExpression(trigger.cronExpression)
+      expr.timeZone = dateContext.timeZone
+
+      if (missedExecution(expr, lastExecution, dateContext.triggerWindowFloor, dateContext.now)) {
+        log.info("Triggering missed execution on pipeline ${pipeline}")
+        pipelineInitiator.call(pipeline)
+      }
+    }
+  }
+
+  void onOrcaError(Throwable error) {
+    log.error("Error retrieving latest pipeline executions", error)
+    counter.increment('orca.errors')
+  }
+
+  static List<Trigger> getEnabledCronTriggers(List<Pipeline> pipelines) {
+    (List<Trigger>) pipelines
+      .collect { it.triggers }
+      .flatten()
+      .findAll { Trigger it -> it && it.enabled && it.type == Trigger.Type.CRON.toString() }
+  }
+
+  static boolean missedExecution(CronExpression expr, Date lastExecution, Date windowFloor, Date now) {
+    // Quartz works at the minute-level, so given 12:00:00, the next "valid" time will be 12:00:01. Without
+    // bumping the lastExecution to the next invalid time, we would double-trigger timely cron executions.
+    lastExecution = expr.getNextInvalidTimeAfter(lastExecution)
+
+    if (lastExecution.before(windowFloor)) {
+      // The last execution is past the point of no return, as far as the compensation job is concerned.
+      return false
+    }
+
+    def nextExecution = expr.getNextValidTimeAfter(lastExecution)
+    while (true) {
+      if (nextExecution.after(windowFloor)) {
+        break
+      }
+      nextExecution = expr.getNextValidTimeAfter(nextExecution)
+    }
+
+    return nextExecution.before(now)
```

Comment:
Maybe worth a comment either here or javadoc on the method indicating that we're looking for any that should have been fired between `windowFloor` and `now` (where `windowFloor` is `now - windowMs` right?)